multimodal/src/autogluon/multimodal/optimization/lit_mmdet.py

```diff
@@ -0,0 +1,321 @@
+import logging
+from typing import Callable, Dict, List, Optional, Union
+
+import numpy as np
+import pytorch_lightning as pl
+import torch
+import torchmetrics
+from omegaconf import DictConfig
+from torch import nn
+from torch.nn.modules.loss import _Loss
+from torchmetrics.aggregation import BaseAggregator
+
+try:
+    import mmdet
+    from mmcv import ConfigDict
+except ImportError:
+    pass
+
+from ..constants import AUTOMM
+from ..utils import unpack_datacontainers
+from .utils import apply_layerwise_lr_decay, apply_single_lr, apply_two_stages_lr, get_lr_scheduler, get_optimizer
+
+logger = logging.getLogger(AUTOMM)
+
+
+class MMDetLitModule(pl.LightningModule):
+    def __init__(
+        self,
+        model,
+        optim_type: Optional[str] = None,
+        lr_choice: Optional[str] = None,
+        lr_schedule: Optional[str] = None,
+        lr: Optional[float] = None,
+        lr_decay: Optional[float] = None,
+        end_lr: Optional[Union[float, int]] = None,
+        lr_mult: Optional[Union[float, int]] = None,
+        weight_decay: Optional[float] = None,
+        warmup_steps: Optional[int] = None,
+        loss_func: Optional[_Loss] = None,
+        validation_metric: Optional[torchmetrics.Metric] = None,
+        validation_metric_name: Optional[str] = None,
+        custom_metric_func: Callable = None,
+        test_metric: Optional[torchmetrics.Metric] = None,
+        efficient_finetune: Optional[str] = None,
+    ):
+        super().__init__()  # TODO: inherit LitModule
+        self.save_hyperparameters(
+            ignore=[
+                "model",
+                "validation_metric",
+                "test_metric",
+                "loss_func",
+                "matches",
+            ]
+        )
+        self.model = model
+        self.validation_metric = validation_metric
+        self.validation_metric_name = f"val_{validation_metric_name}"
+        if isinstance(validation_metric, BaseAggregator) and custom_metric_func is None:
+            raise ValueError(
+                f"validation_metric {validation_metric} is an aggregation metric,"
+                f"which must be used with a customized metric function."
+            )
+        self.custom_metric_func = custom_metric_func
+        self.id2label = self.model.id2label
+
+    def _predict_step(self, batch, batch_idx=0):
+        from mmcv.ops import RoIPool
+        from mmcv.parallel import scatter
+
+        data = batch["mmdet_image_image"]
+        data["img_metas"] = [img_metas.data[0] for img_metas in data["img_metas"]]
+        data["img"] = [img.data[0] for img in data["img"]]
+        if next(self.model.parameters()).is_cuda:
```

Comment:
Why do we need this?

Reply:
The output of collate function is a DataContainer object(s), we need to unpack and send to gpu here.

Reply:
Okay, will data be on CPU if we do not do this? I think this seems like a problem associated with our collate function / the way of integrating the collate function with lightning.

Reply:
Yes it will be on CPU. I'll check if we can move the unpack codes to collate function. Then it will be better integrated with lightning.